### PR TITLE
show type header for serverless function in topology

### PR DESCRIPTION
**Pull request by nemesis09**
Story:
https://issues.redhat.com/browse/ODC-5658

Problem/Description:
The Topology sidebar and Service details page should show that it belongs to serverless function

Solution:
Add a "Type" header with content "Function" under it to indicate the details belong to serverless function

Screens:
_Topology Sidebar for Serverless function_
Screenshot from 2021-03-30 21-22-14

_Service details page for Serverless function_
Screenshot from 2021-03-30 21-23-05

_Topology Sidebar for Knative service_
Screenshot from 2021-03-30 21-22-32

_Service details page for Knative service_
Screenshot from 2021-03-30 21-22-42

_Secret details page_
Screenshot from 2021-04-09 22-34-24

Test Coverage:
Screenshot from 2021-03-30 21-27-00
Screenshot from 2021-03-30 21-30-05

Browser Conformance:
- [x] Firefox
- [ ] Chrome
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Comment by nemesis09** [comment]
/assign @invincibleJai 
cc @openshift/team-devconsole-ux

**Review by bgliwa01 — APPROVED** [review]
LGTM! thanks!

**Comment by invincibleJai** [comment]
> Also I'm not sure if there is another story to take care of this, but there should be a different fx icon in topology. cc @beaumorley

@bgliwa01 yes it handled as part of https://github.com/openshift/console/pull/8525 PTAL , thanks !!

**Comment by nemesis09** [comment]
> > Also I'm not sure if there is another story to take care of this, but there should be a different fx icon in topology. cc @beaumorley
> 
> @bgliwa01 yes it handled as part of #8525 PTAL , thanks !!

Thanks! @invincibleJai

File: frontend/public/components/default-resource.tsx [hunk 1]
```diff
@@ -64,6 +67,9 @@ export const DetailsForKind = (kind: string) =>
                 showNodeSelector={false}
               />
             </div>
+            {additionalDetailsCallback && (
+              <div className="col-md-6">{additionalDetailsCallback(obj)}</div>
```
Comment (sahil143) [inline frontend/public/components/default-resource.tsx:71]: If `additionalDetailsCallback` return null then an empty div will rendered.

File: frontend/public/components/utils/details-page.tsx [hunk 2]
```diff
@@ -91,8 +91,6 @@ export const ResourceSummary: React.FC<ResourceSummaryProps> = ({
           />
         </DetailsItem>
       )}
-      {type ? <dt>{t('details-page~Type')}</dt> : null}
-      {type ? <dd>{type}</dd> : null}
```
Comment (sahil143) [inline frontend/public/components/utils/details-page.tsx:95]: Why is this removed? If any resource had the `type` it would not work anymore

File: frontend/public/components/default-resource.tsx (hunk 1 above)
Comment (nemesis09) [reply frontend/public/components/default-resource.tsx:71]: added check to restrict render when function returns null or undefined.

File: frontend/public/components/utils/details-page.tsx (hunk 2 above)
Comment (nemesis09) [reply frontend/public/components/utils/details-page.tsx:95]: Type was added for "Secrets" alone. No other resources seemed to use this.
any specific type for a resource can be handled using annotations, as I have done for sinkUri.
"type" for Secret has been added to "SecretKind" which is specific to secret.
please refer to this conversation for more information
https://coreos.slack.com/archives/CHGU4P8UU/p1616684514128300

**Review by sahil143 — APPROVED** [review]
/lgtm

verified locally, works as expected

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts [hunk 3]
```diff
@@ -977,9 +978,9 @@ export const transformKnNodeData = (
                 metadata: {
                   uid: sinkTargetUid,
                   namespace: data.resources.obj.metadata.namespace || '',
+                  annotations: { [typeAnnotation]: NodeType.SinkUri },
                 },
                 spec: { sinkUri },
-                type: { nodeType: NodeType.SinkUri },
                 kind: 'URI',
```
Comment (christianvogt) [inline frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts:984]: Can we not use the fake `kind` we created here instead of a new annotation?

File: frontend/packages/knative-plugin/src/components/overview/ServerlessFunctionType.tsx [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import * as React from 'react';
+import { useTranslation } from 'react-i18next';
+
+const ServerlessFunctionType: React.FC = () => {
+  const { t } = useTranslation();
+  return (
+    <dl className="co-m-pane__details">
+      <dt>{t('knative-plugin~Type')}</dt>
+      <dd>{t('knative-plugin~Function')}</dd>
+    </dl>
```
Comment (christianvogt) [inline frontend/packages/knative-plugin/src/components/overview/ServerlessFunctionType.tsx:10]: Shouldn't we be using the `DetailsItem` component?
```
<DetailsItem label={t('knative-plugin~Type')}>{t('knative-plugin~Function')}</DetailsItem>
```

File: frontend/public/components/utils/details-page.tsx (hunk 2 above)
Comment (christianvogt) [reply frontend/public/components/utils/details-page.tsx:95]: You've removed this from the details view. However I do not see that you've added it back in the right column for secrets

**Comment by nemesis09** [comment]
> You've removed this from the details view. However I do not see that you've added it back in the right column for secrets
This wasn't being used by any component.

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts (hunk 3 above)
Comment (nemesis09) [reply frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts]: The concern was to keep it specific to SinkUri. So we chose to use type `sink-uri` and annotaion seemed to be the proper place to use it.
cc @invincibleJai

File: frontend/packages/knative-plugin/src/topology/const.ts [hunk 5]
```diff
@@ -19,3 +19,6 @@ export const KNATIVE_GROUP_NODE_PADDING = [
 ];
 
 export const EVENT_MARKER_RADIUS = 6;
+
+// Annotation to identify a specific type of resource bearing the same resource "Kind"
+export const typeAnnotation = 'knative.openshift.io/type';
```
[the remark this answers is not in the document]
Comment (nemesis09) [reply frontend/packages/knative-plugin/src/topology/const.ts]: done

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts (hunk 3 above)
Comment (invincibleJai) [reply frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts]: @nemesis09 we can go with `kind` as suggested, second thought was if in future any resource has kind `URI` which is not the case so far

Comment (nemesis09) [reply frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts]: updated to use kind

File: frontend/public/components/utils/details-page.tsx (hunk 2 above)
Comment (nemesis09) [reply frontend/public/components/utils/details-page.tsx:95]: this was never used by any component. Using this component would result in a bug being observed since it was rendering an object.
`SecretDetails` already receives a `secret.type`. I just updated it to use `SecretKind` instead of `K8sResourceKind`.

https://github.com/openshift/console/blob/dc102a74f4bbd98607c969b67fa13e1f31dbc6b1/frontend/public/components/secret.jsx#L116

**Comment by invincibleJai** [comment]
> > You've removed this from the details view. However I do not see that you've added it back in the right column for secrets
> 
> Apparently, this wasn't being used by any component.

@nemesis09 it was shown in details look like see below 

image

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts (hunk 3 above)
Comment (christianvogt) [reply frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts]: Since this is a made up resource, we could make a more unique kind as well and even pair it with a unique api group.

File: frontend/packages/knative-plugin/src/components/overview/ServerlessFunctionType.tsx (hunk 4 above)
Comment (christianvogt) [reply frontend/packages/knative-plugin/src/components/overview/ServerlessFunctionType.tsx:10]: Ok that's fine.

**Comment by nemesis09** [comment]
added type for secret details page
Thanks for bringing it to notice @invincibleJai @christianvogt

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts [hunk 6]
```diff
@@ -1140,7 +1140,7 @@ export const createKnativeEventSourceSink = (
   }
   const eventSourceObj = _.omit(source, 'status');
   let sink = {};
-  if (NodeType.SinkUri === target.type?.nodeType) {
+  if (NodeType.SinkUri === target.kind) {
```
Comment (christianvogt) [inline frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts:1143]: `NodeType.SinkUri` has a value of `'sink-uri'`.

```suggestion
  if (URI_KIND === target.kind) {
```

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts [hunk 7]
```diff
@@ -1246,13 +1246,15 @@ export const createSinkPubSubConnection = (
   return createEventingPubSubSink(resources.obj, targetObj);
 };
 
-export const isServerlessFunction = (element: Node): boolean => {
+export const isServerlessFunction = (element: Node | K8sResourceKind): boolean => {
```
Comment (christianvogt) [inline frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts:1249]: Better to create two separate functions or change it to this.
```suggestion
export const isServerlessFunction = (resource: K8sResourceKind): boolean => {
```

Then when you have a topology node in hand, always call it like this `isServerlessFunction(getResource(node))`

File: frontend/public/components/secret.jsx [hunk 8]
```diff
@@ -110,10 +111,18 @@ const SecretDetails = ({ obj: secret }) => {
           <div className="col-md-6">
             <ResourceSummary resource={secret} />
           </div>
+          {type && (
+            <div className="col-md-6">
+              <dl className="co-m-pane__details">
+                <dt>{t('public~Type')}</dt>
+                <dd>{type}</dd>
```
Comment (christianvogt) [inline frontend/public/components/secret.jsx:118]: Shouldn't this be using `DetailsItem` since it's value directly from the resource? `DetailsItem` does extra work to potentially pull out descriptions from metadata.
```suggestion
                <DetailsItem label={t('public~Type')} obj={secret} path="type" />
```

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts (hunk 6 above)
Comment (nemesis09) [reply frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts]: done

File: frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts (hunk 7 above)
Comment (nemesis09) [reply frontend/packages/knative-plugin/src/topology/knative-topology-utils.ts]: done

File: frontend/public/components/secret.jsx (hunk 8 above)
Comment (nemesis09) [reply frontend/public/components/secret.jsx]: done

Comment (christianvogt) [reply frontend/public/components/secret.jsx]: Nice!
image

**Comment by christianvogt** [comment]
/hold
consistent failure of cypress test on accessibility error

**Comment by nemesis09** [comment]
pushed changes to address failing cypress tests
/hold cancel
